### PR TITLE
Use reflection provider for class constructors

### PR DESCRIPTION
**Pull request by nohwnd**
Fixes oversight in code where we grab the constructors from Reflection rather than getting them from the metadata provider.

Replace https://github.com/microsoft/testfx/pull/4033

### CODE REVIEW

**Review by Evangelink — DISMISSED** [review]

**Review by Evangelink — COMMENTED** [review]
Some tests are now breaking on `MSTestAdapter.UnitTest`

File: src/Adapter/MSTest.TestAdapter/Execution/TypeCache.cs [hunk 1]
```diff
@@ -260,24 +260,42 @@ private static bool TryGetUnescapedManagedTypeName(TestMethod testMethod, [NotNu
...
+            }
+
+            ParameterInfo[] parameters = ctor.GetParameters();
+
+            // There are just 2 ctor shapes that we know, so the code is quite simple,
+            // but if we add more, add a priority to the search, and shortcircuit this search so we only iterate
+            // through the collection once, to avoid re-allocating GetParameters multiple times.
+            if (parameters.Length == 1 && parameters[0].ParameterType == typeof(TestContext))
+            {
+                selectedConstructor ??= (ctor, IsParameterless: false);
```
Comment (Youssef1313) [inline src/Adapter/MSTest.TestAdapter/Execution/TypeCache.cs]: If we first found a parameterless constructor, then we found this, we want to always assign right?

```suggestion
                selectedConstructor = (ctor, IsParameterless: false);
```

File: src/Adapter/MSTest.TestAdapter/Execution/TypeCache.cs [hunk 2]
```diff
@@ -260,24 +260,42 @@ private static bool TryGetUnescapedManagedTypeName(TestMethod testMethod, [NotNu
...
+
+                // This is the preferred constructor, no point in searching for more.
+                break;
+            }
+
+            if (parameters.Length == 0)
+            {
+                selectedConstructor = (ctor, IsParameterless: true);
```
Comment (Youssef1313) [inline src/Adapter/MSTest.TestAdapter/Execution/TypeCache.cs]: Maybe add `Debug.Assert(selectorConstructor is null);` before this line?

File: src/Adapter/MSTest.TestAdapter/Execution/TypeCache.cs (hunk 1 above)
Comment (nohwnd) [reply src/Adapter/MSTest.TestAdapter/Execution/TypeCache.cs]: I think I meant to put this on the other case, so we select the first parameterless one.

File: src/Adapter/MSTest.TestAdapter/Execution/TypeCache.cs (hunk 2 above)
Comment (nohwnd) [reply src/Adapter/MSTest.TestAdapter/Execution/TypeCache.cs]: moving the `??=` to the correct place solves the issue that was motivation for this Assert. Which was to error when we forget to break in the most preferred case, then falling through here would choose the non-preferred constructor.

**Pull request merged**